### ISSUE TITLE
CSS resolving order

### ISSUE THREAD

**Issue by gunta** [webpack/webpack#215]
Hi!

I have two files

``` javascript
require('bootstrap.css')
require('./myfile.css')
```

where the file `bootstrap.css` **needs** to be loaded before `./myfile.css`.
Is there any way to do set the CSS resolving order properly in webpack for CSS?

**Comment by crittelmeyer** [comment]
+1

**Comment by luisrudge** [comment]
This still happens. What's the workaround?

**Comment by sokra** [comment]
@luisrudge Check if you have the lastest version of webpack, extract-text-webpack-plugin and css-loader. This shouldn't happen.

**Comment by luisrudge** [comment]
I have, but it still happens.

**Comment by raypatterson** [comment]
I have observed this issue using bot a legacy and (at the time of writing) most up-to-date webpack and loaders. It appears to only happen on on rebuilds, however. 

For example, this works as expected: 
```
require('normalize.css');
require('app.css');
```

But if I comment out the 2nd line, save, wait for rebuild, uncomment, save, wait for rebuild, the order of the file content in the output is reversed.

I am splitting code with the CommonChunksPlugin but haven't explored if that matters.

**Comment by raypatterson** [comment]
Finding this only happens on rebuilds.

My workaround it to have an entry point for each "module" that has SASS dependencies.

```
/* my-module/index.js */

require('./index.sass');

/* my-module/index.sass */

@import 'sass-dependency-a';
@import 'sass-dependency-b';
```

**Comment by jcreamer898** [comment]
I'm seeing a very similar issue. Going to try the entry hack that @raypatterson is trying too.

**Comment by timaschew** [comment]
I have this issue only if I use the [extract-text-webpack-plugin](https://github.com/webpack/extract-text-webpack-plugin).
If I turn it off the order is correct

**Comment by sokra** [comment]
@timaschew Have you tried the latest versions for css-loader extract-text-webpack-plugin and webpack? They should have the correct css order.

**Comment by timaschew** [comment]
I've updated now from css-loader __0.14.5__ to __0.15.1__, the extract plugin is up to date at __0.8.2__
 but unfortunately it still doesn't work.

It's normalize.css which is almost at the end of the css file.
I used some console.logs before requiring the styles and in the browser console, the logs are in the right order

**Comment by timaschew** [comment]
I can reproduce the issue in this repo: https://github.com/timaschew/webpack-amd-commonJs-demo

**Comment by timaschew** [comment]
I've started to debug the issue
So first of all, the [extractedChunks](https://github.com/webpack/extract-text-webpack-plugin/blob/a03c82c59d557ec4abe1f9afff8ad7cb91325eb0/index.js#L209) have the wrong order from the beginning.

But there is an [getOrder](https://github.com/webpack/extract-text-webpack-plugin/blob/a03c82c59d557ec4abe1f9afff8ad7cb91325eb0/index.js#L89) function which is comparing against the `index2` property.

Then I printed out the both items which are compared
````
{ _identifier: '/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/css-loader/index.js!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/normalize.css/normalize.css',
  _originalModule: 
   { dependencies: [],
     blocks: [],
     variables: [],
     context: '/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/normalize.css',
     reasons: [ [Object] ],
     debugId: 1004,
     lastId: -1,
     id: 9,
     index: 7,
     index2: 6,
     chunks: [ [Object] ],
     warnings: [],
     dependenciesWarnings: [],
     errors: [],
     dependenciesErrors: [],
     request: '/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/extract-text-webpack-plugin/loader.js?{"omit":1,"extract":true,"remove":true}!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/style-loader/index.js!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/css-loader/index.js!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/normalize.css/normalize.css',
     userRequest: '/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/normalize.css/normalize.css',
     rawRequest: 'normalize.css/normalize.css',
```
and
```
{ _identifier: '/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/css-loader/index.js!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/stylus-loader/index.js?{"resolve url": true}!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a/styles.styl',
  _originalModule: 
   { dependencies: [],
     blocks: [],
     variables: [],
     context: '/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a',
     reasons: [ [Object] ],
     debugId: 1006,
     lastId: -1,
     id: 4,
     index: 2,
     index2: 4,
     chunks: [ [Object] ],
     warnings: [],
     dependenciesWarnings: [],
     errors: [],
     dependenciesErrors: [],
     request: '/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/extract-text-webpack-plugin/loader.js?{"omit":1,"extract":true,"remove":true}!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/style-loader/index.js!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/css-loader/index.js!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/stylus-loader/index.js?{"resolve url": true}!/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a/styles.styl',
     userRequest: '/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a/styles.styl',
     rawRequest: './styles.styl',
```

So, okay the order function is working, but why the values for `index2` have the wrong order?
So I found that webpack has two places where this variable is set
1. [lib/Compilation.js#seal](https://github.com/webpack/webpack/blob/ba03fda4b5d2a515ccd87d26f3edd0dd357c4fde/lib/Compilation.js#L504)
2. [[lib/Compilation.js#iteratorDependency](https://github.com/webpack/webpack/blob/ba03fda4b5d2a515ccd87d26f3edd0dd357c4fde/lib/Compilation.js#L643)

If I add some logs between these lines and combine some others from the css-loader and extract-plugin I get this:

```
>>>> css-loader/loader
>> normalize.css
>>>> css-loader/loader
>> styles.styl
####l lib/Compilation.js#seal
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/css-loader/lib
##i index2: 0
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-b/icons
##i index2: 1
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a
##i index2: 2
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/style-loader
##i index2: 3
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a
##i index2: 4
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/normalize.css
##i index2: 5
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/normalize.css
##i index2: 6
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/domify
##i index2: 7
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-b
##i index2: 8
####i lib/Compilation.js#iteratorDependency
##i null
##i index2: 9
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/jade/lib
##i index2: 10
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a
##i index2: 11
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-b
##i index2: 12
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a
##i index2: 13
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a
##i index2: 14
##l /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src
##l index2: 15
<<<< extract plugin
>>>> css-loader/loader
>> normalize.css
>>>> css-loader/loader
>> styles.styl
####l lib/Compilation.js#seal
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/css-loader/lib
##i index2: 0
##l /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/normalize.css
##l index2: 1
####l lib/Compilation.js#seal
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/css-loader/lib
##i index2: 0
####i lib/Compilation.js#iteratorDependency
##i /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-b/icons
##i index2: 1
##l /Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a
##l index2: 2
<< getOrder, index2
/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/node_modules/normalize.css/normalize.css
6
---
/Users/awilhelm/dev/webpack-test/webpack-amd-commonJs-demo/src/feature-a/styles.styl
4
```

At this point it's too complex for me ^^
I don't get it, how and why `index2` is set and why for normalize it's called two times.

**Comment by vn38minhtran** [comment]
Still happen with `webpack 1.10.1`, `css-loader 0.15.2`, `extract-text-webpack-plugin 0.8.2`.
- app.js
```js
...
require('./bower_components/bootstrap/css/bootstrap.css');
require('./assets/styles/app.scss');
...
```
- app.scss
```css
body {
  background: red;
}
```
- result:
![screenshot from 2015-07-12 23 34 25](https://cloud.githubusercontent.com/assets/5418628/8638730/952dc13c-28ee-11e5-8d8b-8ca4f93510c3.png)

**Comment by jackmoore** [comment]
Having the same problem as @timaschew and @vn38minhtran

**Comment by Coobaha** [comment]
+1

**Comment by luisrudge** [comment]
any thouhgts on this?

**Comment by jackmoore** [comment]
@luisrudge @Coobaha I just bumped my installed webpack version to 1.11.0 and css-loader to 0.15.6 and I started getting the correct import order, so this may be fixed.

**Comment by luisrudge** [comment]
I'll try that

**Comment by luisrudge** [comment]
Still doesn't work for me :(

**Comment by luisrudge** [comment]
![image](https://cloud.githubusercontent.com/assets/941075/9117702/49df24f2-3c42-11e5-8924-ee8bcd962314.png)

**Comment by jcreamer898** [comment]
Still seeing the issue here too.

**Comment by jcreamer898** [comment]
These font rules are supposed to be overridden for me, they are being put at the bottom of my `common.css` output... 

![](http://d.pr/i/1cQDN+)

I'm using Extract + CommonChunks.

**Comment by newtonianb** [comment]
I'm having the same issue. I tried updating to webpack 1.11 and css loader but it made no difference. My bootstrap style is being loaded before my custom style.

I have two entry points
```
entry.js
    require('bootstrap/dist/css/bootstrap.min.css')
    require('bootstrap/dist/js/bootstrap.min.js')
```
```
app.js
   require('mystyle.scss')
```
```
index.html
      vendor.js
      app.js
```

**Comment by hank7444** [comment]
+1

**Comment by garretto** [comment]
+1

**Comment by adrianca** [comment]
Also having this issue :+1:

**Comment by umidbekkarimov** [comment]
+1

**Comment by cdomergue** [comment]
+1

**Comment by sokra** [comment]
As workaround require the vendor entry point in the application entry point.

**Comment by luisrudge** [comment]
@sokra what do you mean? I'm already requiring bootstrap's main file from my entry point file. Do I need to do something different?

**Comment by sokra** [comment]
The problem is propably the CommonChunkPlugin. webpack doesn't "learn" order from it. From webpacks point it just a bunch of entry points without order. So you can give webpack a hint by adding a dependency from one entry point to another (The CommonsChunkPlugin will optimize it away).

**Comment by luisrudge** [comment]
I'm not using `CommonChunkPlugin`

**Comment by luisrudge** [comment]
I'm using this plugins:
```
new webpack.optimize.OccurrenceOrderPlugin()
new webpack.optimize.DedupePlugin()
new ExtractTextPlugin('bundle.css')
```

**Comment by sokra** [comment]
A single entry or multiple?

**Comment by luisrudge** [comment]
Single.

`entry: { app: [ './App/app.js' ] }`

**Comment by sokra** [comment]
Is extract-text-webpack-plugin, css-loader and webpack up-to-date?

**Comment by luisrudge** [comment]
yes.

```
"extract-text-webpack-plugin": "0.8.2",
"css-loader": "0.19.0",
"webpack": "1.12.2",
```

**Comment by luisrudge** [comment]
My app.js file has this: `import 'app.less';`
My app.less file has this: `@import 'CustomBootstrap.less';`
My CustomBootstrap.less file has this:
```css
@import '../node_modules/bootstrap/less/bootstrap.less';
@import 'CustomBootstrap.Variables.less';
@import 'CustomBootstrap.Styles.less';
```

**Comment by sokra** [comment]
hmm... Could you make a repo with the bug?

**Comment by luisrudge** [comment]
I'll try to create a repo with the bug in the weekend

**Comment by sokra** [comment]
Thanks

**Comment by dready** [comment]
any news? same problem here, no CommonChunkPlugin and i need the extract text webpack plugin.

latest webpack and loaders

**Comment by ponelat** [comment]
+1

**Comment by kpeu3i** [comment]
+1

**Comment by andykant** [comment]
I've been running into this issue as well, but can't track down a use case of what's causing it. It seems inconsistent, but OccurrenceOrderPlugin helps make it more consistent.

**Comment by samloh84** [comment]
+1

**Comment by cdomergue** [comment]
I've been running into this issue as well, for jQuery and bootstrap.

I resolved this by adding jQuery to plugin in webpack config :

```
new webpack.ProvidePlugin({
            jQuery: 'jquery',
            $: 'jquery
```

**Comment by fabm22** [comment]
Hi, any news on this  issue ?

Also, has anyone tried to use require.ensure  to specify the desired order ?

```
 var a = require("a");
 require.ensure(["b"], function(require) {
    require("a").dosomething();
    var c = require("c");
});
```

**Comment by MartinMystikJonas** [comment]
+1

**Comment by krazi3** [comment]
+1

**Comment by bartko-s** [comment]
+1

**Comment by studiosciences** [comment]
1) In my tests without the Extract Text Plugin, load order is very reliable and predictable, even with Live Reload and a hierarchy of views each with with their own css. The first instance of a css file is loaded first.

2) What's not clear to me is why the [Webpack documentation](https://webpack.github.io/docs/stylesheets.html) says:

> Keep in mind that it’s difficult to manage the execution order of modules, so design your stylesheets so that order doesn’t matter. (But you can rely on the order in one css file.)

Is this no longer the case? Or does this just mean don't create undefined relationships between css files and js views?  It seems that require(css) as an architecture wouldn't work if load order couldn't be controlled.

3) Given that no one has provided reproduction steps and many of the comments are off topic, I really hope this is just individuals' misconfigurations or a misalignment of css and js views.

**Comment by l0gicgate** [comment]
`compilation.plugin("additional-assets", function(callback) {
                        extractedChunks.forEach(function(extractedChunk) {
                                if(extractedChunk.modules.length) {
                                        /* This is the culprit right here, simply comment this out, this is a temporary fix obviously */
                                        /* extractedChunk.modules.sort(function(a, b) {
                                                if(isInvalidOrder(a, b)) {
                                                        compilation.errors.push(new OrderUndefinedError(a.getOriginalModule()));
                                                        compilation.errors.push(new OrderUndefinedError(b.getOriginalModule()));
                                                }
                                                return getOrder(a, b);
                                        }); */
                                        var chunk = extractedChunk.originalChunk;
                                        var source = this.renderExtractedChunk(extractedChunk);
                                        var file = compilation.getPath(filename, {
                                                chunk: chunk
                                        }).replace(/\[(?:(\w+):)?contenthash(?::([a-z]+\d*))?(?::(\d+))?\]/ig, function() {
                                                return loaderUtils.getHashDigest(source.source(), arguments[1], arguments[2], parseInt(arguments[3], 10));
                                        });
                                        compilation.assets[file] = source;
                                        chunk.files.push(file);
                                }
                        }, this);
                        callback();
                }.bind(this));
`

**Comment by l0gicgate** [comment]
In "node_modules/extract-text-webpack-plugin/index.js" comment out the following code for a temporary fix:


     /*
     extractedChunk.modules.sort(function(a, b) {
          if(isInvalidOrder(a, b)) {
               compilation.errors.push(new OrderUndefinedError(a.getOriginalModule()));
               compilation.errors.push(new OrderUndefinedError(b.getOriginalModule()));
          }
          return getOrder(a, b); 
     }); 
     */

**Comment by epilande** [comment]
+1 

Is there any way to set load order?

**Comment by srahulprdxn** [comment]
This problem still exists.

@import '../node_modules/bootstrap/bootstrap.min.css';
@import 'reset.scss';

{
        'test': [/\.css$/, /\.scss$/],
        'loader': 'style!css!sass'
}

Reset loads first.

**Comment by reverland** [comment]
just have different production effect with webpack dev server, and find this issue...

**Comment by mcbobo** [comment]
+1

Not sure if this is the same issue. I'm using react / webpack / sass.

I'm trying to have a main layout component and have a `core.scss`, which has a partial imported, `_base.scss`, which imports all my `variables` such as colors.

But then when I try to include components such as a `Nav` react component, and it has it's own component CSS such as `Nav.scss`, it doesn't have access to the colors since the `Nav.scss` loaded before the `core.scss`. Tested this with modifying the body colors and it indeed does load afterwards...

We should definitely build in support to change the load order. My current work around is to re-import the colors file.

**Comment by Andrey-Pavlov** [comment]
Any workaround for this bug?

**Comment by srahulprdxn** [comment]
@Andrey-Pavlov For now, I embed all the base css code like reset.css or main.css in the index file itself.

**Comment by billyxs** [comment]
I have this issue when using bootstrap loader.  

```webpackConfig.entry = {
  app: __DEV__
    ? [APP_ENTRY_PATH, 'bootstrap-loader',
      `webpack-hot-middleware/client?path=${config.compiler_public_path}__webpack_hmr`]
    : [APP_ENTRY_PATH, 'bootstrap-loader'],
  vendor: config.compiler_vendor
}```

With this config, I get normalize and bootstrap loaded after my css
![bootstrap-loader](https://cloud.githubusercontent.com/assets/404063/15980897/d3ec4a74-2f23-11e6-9567-6a31465b4cab.png)

**Comment by billyxs** [comment]
I have this issue when using bootstrap loader.  

```
webpackConfig.entry = {
  app: __DEV__
    ? [APP_ENTRY_PATH, 'bootstrap-loader',
      `webpack-hot-middleware/client?path=${config.compiler_public_path}__webpack_hmr`]
    : [APP_ENTRY_PATH, 'bootstrap-loader'],
  vendor: config.compiler_vendor
}
```
With this config, I get normalize and bootstrap loaded after my css in the `<style>` tag append. Not sure how to control the order

![bootstrap-loader](https://cloud.githubusercontent.com/assets/404063/15980983/c2584d0c-2f24-11e6-9dad-84a4ee5eab9a.png)

**Comment by shalanah** [comment]
+1

**Comment by nuragic** [comment]
Hi all,

I got the same problem, but I also find the solution (well, it worked for me... hopefully could be helpful for someone else). 

In the middle of my codebase, there was a **dynamic require**, something like:
```
this.loadImage = (path) => require(`../../../../${path}`)
```

Ok so seems that this 💩y line was tricking webpack in some way **_to load every single file in the code base_**... killer! 😲  Removing that line solved this weird issue (and some other strange warning message in the console that started appearing...oh my god).

Dynamic requires... eeeh! 💀

**Comment by andreechristian** [comment]
+1

**Comment by hph** [comment]
This is still a problem and I'm not sure why, however, in my case it had a simple fix. I saw this issue when a `main` tag that I had set to `display: flex` was instead being set to `display: block`, because of the defaults from `normalize.css`, which are incorrectly not at the top of the stylesheet.

In my case, all I had to do (which I should've been doing in the first place) was to use a classname to style the tag instead. So (correct me if I'm wrong) but I think that most of the issues arising from weird ordering of stylesheets should be solvable by strictly enforcing something like BEM or CSS modules.

**Comment by Jwashton** [comment]
Um... @hph I get your point, and I'm glad to hear that the solution was easy to solve in your case (and if applied widely, in most cases.) However, since I typically push back against using non-semantic or redundant classes (because I'm religious about all the wrong things), and typically don't use any specific module implementation other than the native Shadow DOM (yes, I live in the future), I'm afraid enforcing BEM or CSS modules wouldn't really fly well here.

Webpack needs to get the ordering right without forcing us to change the way we think about solving problems.

**Comment by hph** [comment]
@Jwashton I completely agree that this should be fixed and I'm not suggesting otherwise, merely mentioning an acceptable workaround for many people while this is still an issue.

**Comment by Kovensky** [comment]
I haven't tried using CSS with webpack yet, but from what I can tell on the code snippets in here, it looks like dependencies are underspecified.

You are requiring the files in a particular order _at runtime_, but from a dependency tracker's point of view, you're just a file that depends on an _unordered set_ of different files.

If your `index.css` file needs a dependency to appear before it, then the `index.css` itself should `@import` the dependency, telling the dependency resolver that the dependency should come first.

The other snippet with an `index.`~~`js`~~`sass` just moves the dependency problem down a level; the dependency tracker still doesn't know that `reset.scss` depends on `bootstrap.min.css`. It just happens that webpack does "execute" the sass before extracting the css, so the "runtime" order is preserved.

That is, assuming this is not actually an extract-text-plugin or common-chunks-plugin bug.

**Comment by schneidmaster** [comment]
I suspect this is an issue in the extract-text-webpack-plugin and/or the css-loader. I worked around it by adding the `-minimize` flag to the css-loader as suggested [here](https://github.com/webpack/extract-text-webpack-plugin/issues/171#issuecomment-224618774). As you would suspect, this also turns off CSS minification in production; I added [optimize-css-assets-webpack-plugin](https://www.npmjs.com/package/optimize-css-assets-webpack-plugin) to get minified CSS again.

**Comment by omarjcamo** [comment]
+1

**Comment by srahulprdxn** [comment]
+1

**Comment by jverrone3** [comment]
+1

**Comment by tugberkugurlu** [comment]
I am also hitting this. Did anyone find a solution or workaround?

**Comment by cage1618** [comment]
+1

**Comment by reverland** [comment]
People who just +1 may should check their webpack configs. There are some ideas may help

1. try to enable/disable the loaders/plugins
2. import will always be hoist but require wont
3. do you extract styles from `node_modules` and then use common chunks plugin to merge it into a single file, and then when it is insert into the html file in the unwanted order?
4. do you async load styles? and style loaded order is really you want?
5. and so on

I hope this will be help, for webpack is quite complex but versatile,  many may don't know what happened under the tons of configs in minutes.

`+1` may not help at all IMHO

**Comment by reverland** [comment]
I always feel confused with styles 

is there a good way for us to handle styles in webpack, css may not well designed to fit in javascript modules.

or we should just use something like less/sass/stylus and so on,?

or mysterious css in js or css module?

**Comment by ponelat** [comment]
I kind of feel the same.
We use `less + common chunks plugin`, but the weird order of the css ( I say weird, as I don't fully understand it... and if something breaks I just jiggle it until it works ).

**Comment by pavelkornev** [comment]
@andreechristian

> +1 This is really bugging me. In the end i need to prepend my base css into my webpack build using gulp :(

Wait a minute. Why don't you just include your base css in your entry point at the very beginning (before importing other components)?

**Comment by Restuta** [comment]
Sokra asked @luisrudge to create a repo  that reproduces this bug, he never did. I suppose this is a good next step for everybody who is affected. Fairy minor contribution, but at least can help maintainer to put this issue into the queue. Much more helpful than +1's or workarounds.

@cage1618 @tugberkugurlu @jverrone3 @srahulprdxn @omarjcamo @shalanah @epilande

**Comment by cage1618** [comment]
@Restuta Thanks for reminding, I will do this  in this weekend.

**Comment by andreechristian** [comment]
@pavelkornev i was pretty sure it is not going to solve it. But then i tried and then it works! I guess, my mistake was to include the base css on Page container level. Now i include it on the page controller level. The ordering works now. Thanks!

**Comment by kevinSuttle** [comment]
I don't use the extract-text plugin and I put the CSS at the top of my main JS entrypoint. Still out of order.

**Comment by kevinSuttle** [comment]
I wonder if using a Map would help. 
https://github.com/MoOx/postcss-cssnext/issues/282

**Comment by AoDev** [comment]
Hi, in my case the order can be ok or not. I use the less preprocessor and different loaders depending on when I want to speed up things.

__Configuration when the order is ok (slower)__
Here we have sourcemaps and css cleanup.
```
let lessLoader = {
  test: /\.less$/,
  loader: ExtractTextPlugin.extract('style-loader', 'css-loader?sourceMap!less-loader?sourceMap'),
  lessPlugins: [
    new LessPluginCleanCSS({
      advanced: true
    })
  ]
}
```

__Configuration when it's NOT ok__ (faster)
No sourcemap, no CSS cleanup, for local development.
```
if (fasterLess) {
  lessLoader = {
    test: /\.less$/,
    loader: 'style!css!less'
  }
}
```

I don't have time to investigate myself but this can provide a hint to someone...

**Comment by trsh** [comment]
+1

**Comment by 304NotModified** [comment]
+1

**Comment by axelboc** [comment]
Same problem here, specifically when combining nested SCSS imports and webpack `node_modules` imports (with `~` prefix).

**index.scss**
```
@import "comp-1";
@import "~external-comp";
@import "comp-2";
```

**Expected output**
```
/* comp-1 */
/* external-comp */
/* comp-2 */
```

**Actual output**
```
/* external-comp */
/* comp-1 */
/* comp-2 */
```

This occurs whether I use `ExtractTextPlugin` or not. My SCSS loader config is as follows:

```
{
  test: /\.scss$/,
  loader: 'style!css?sourceMap&importLoaders=2!resolve-url!sass?sourceMap'
}
```

`importsLoaders doesn't seem to have an effect either, as the issue remains when I remove it. The `node_modules` import always ends up at the top of the CSS output...

**Comment by montogeek** [comment]
Old code:
```js

**Comment by PeterStegnar** [comment]
Current solution I have found (using SASS to be specific) is to put dependencies to the sass file itself. 

So for example that we have problem with order like:

```
module.js
   include a.css
   include b.scss
```

in index.html we got wrong order like:
```
   style b.css
   style a.css
```

All we need to do is to define depedencies explicitly in a scss file, like in b.css:
 
`@import 'a.css';`

Then webpack can properly resolve dependencies as they are explicitly defined.

**Comment by toothGandalf** [comment]
Wow what a tool! 
And so many great things to discover in webpack for a newcomers (as myself).
I should write another article about how cool webpack is!
> Mar 28, 2014 — CSS resolving order

**Comment by ctrlaltdylan** [comment]
@PeterStegnar Good idea, but didn't work for my setup for whatever reason. Going back to the drawing board on this one

**Comment by secondstreetmedia** [comment]
So, I was having this same problem, but I stumbled across something that, at least initially, appears to be working. The way I originally had my loaders set up for handling styling extensions (.css, .scss, etc.) was this: 



      {
        test: /\.css(\.erb)?$/,
        use: ExtractTextPlugin.extract({
              fallback: 'style-loader',
              use: ['css-loader', 'r-erb-loader']
        })
      },

      {
        test: /\.*(sass|scss)$/,
        use: ExtractTextPlugin.extract({
          fallback: 'style-loader',
          use: ['css-loader', 'sass-loader']
        })
      },

**Comment by 304NotModified** [comment]
I don't mind that there are issues which are open for years.

But it's unclear to me if this will be ever fixed. @sokra please your opinion and plans!

**Comment by secondstreetmedia** [comment]
**tldr;** If you are currently using **`extract-text-webpack-plugin`** for only ***some*** of your styling related loaders (css, scss, sass), try using it for ***all*** of them.

If your loaders for handling styling extensions (.css, .scss, etc.) are set up like mine were, than here is solution which appears to be working for me and may work for you. In my original set up, I was using the **`extract-text-webpack-plugin`** for my **`.scss`** and **`.sass`** files, but ***not*** for my **`.css`** files, which are likely most, if not all, of the stylesheets from the node modules I am using (e.g. tether, bootstrap, etc.):

**config** (original)

    {
      test: /\.css$/,
      use: ['style-loader', 'css-loader']
    },
    {
      test: /\.*(sass|scss)$/,
      use: ExtractTextPlugin.extract({
        fallback: 'style-loader',
        use: ['css-loader', 'sass-loader']
      })
    },

**requires** (using aliases, in the order I want them to load)

    require('tetherCss');
    require('tetherThemeArrowsCss');
    require('tetherThemeArrowsDarkCss');
    require('tetherThemeBasicCss');
    require('bootstrapCss');
    require('materialCss');
    require('materialProjectCss');
    require('pygmentsRougeDefaultCss');
    require('globalStyleScss');

But the result was that none of my custom styling from **`globalStyleScss`** was having any affect. Additionally, all of the styles from the other stylesheets were each appearing in there own separate set of style tags. Not necessarily a problem, but different than what you'll see below. They were, however, appearing in the order I required them.

**index.html** (after webpack)

    <html lang='en'>
      <head>
        <meta content='text/html; charset=UTF-8' http-equiv='Content-Type'>
        <meta charset='utf-8'>
        <meta content='width=device-width, initial-scale=1, shrink-to-fit=no' name='viewport'>
        <title>My App</title>
        <link rel="stylesheet" media="screen" href="http://localhost:8080/global.css" />
        <script src="http://localhost:8080/global.js"></script>
        <style type="text/css">...</style>    
        <style type="text/css">...</style>    
        <style type="text/css">...</style>    
        <style type="text/css">...</style>    
        <style type="text/css">...</style>    
        <style type="text/css">...</style>    
        <style type="text/css">...</style>    
        <style type="text/css">...</style>    
      </head>

It occurred to me to try using the **`extract-text-webpack-plugin`** for ***all*** styling files, then I would at least know they were all being handled in the same way. So, I changed my configuration to this:

**config** (revised)

    {
      test: /\.css$/,
      use: ExtractTextPlugin.extract({
        fallback: 'style-loader',
        use: ['css-loader']
      })
    },
    {
      test: /\.*(sass|scss)$/,
      use: ExtractTextPlugin.extract({
        fallback: 'style-loader',
        use: ['css-loader', 'sass-loader']
      })
    },

Suddenly, I had my custom styling. And notice now, there are no longer multiple style tags for styles from each of the stylesheets. There is only one stylesheet link tag, which combines all of the styles into one, in the order I listed them, with my custom styles at the bottom of all others, just like I want them.

**index.html** (after webpack, revised)

    <html lang='en'>
      <head>
        <meta content='text/html; charset=UTF-8' http-equiv='Content-Type'>
        <meta charset='utf-8'>
        <meta content='width=device-width, initial-scale=1, shrink-to-fit=no' name='viewport'>
        <title>My App</title>
        <link rel="stylesheet" media="screen" href="http://localhost:8080/global.css" />
        <script src="http://localhost:8080/global.js"></script>
      </head>

Just to be sure, I created another custom stylesheet **`globalExtraStyleScss`**, which had styling which would override **`globalStyleScss`**, and I required it ***after*** **`globalStyleScss`**. Reloading the page, I was indeed getting the styling from **`globalExtraStyleScss`**. When I moved the require for **`globalExtraStyleScss`** ***before*** **`globalStyleScss`** and reloaded the page, I got the styling from **`globalStyleScss`** as desired.

I haven't done extensive testing to see if this has permanently solved the issue for me, but I've done enough switching back and forth to see that it likely isn't random. Perhaps others can try it out and report their results. If it should suddenly stop working for me, I will report back.

Hope it helps. Cheers!

**Comment by eladchen** [comment]
@secondstreetmedia This isn't a viable solution. Usually extract-text-webpack-plugin is used in production builds only, As the plugin doesn't support hot-module-replacement.

Using your proposition, during dev build, this issue will occur without using the plugin.

**Comment by timse** [comment]
does anyone feel like producing a small repo to reproduce this problem?:)

**Comment by tpraxl** [comment]
+1 for fixing this

**Comment by kevinarrieta** [comment]
+1

**Comment by ThibaultJanBeyer** [comment]
+1

**Comment by winnerhp** [comment]
+1 please fix it

**Comment by EnricoVogt** [comment]
+1

**Comment by Puerta121** [comment]
+1

**Comment by 304NotModified** [comment]
Any update/plans from the Webpack team? Why isn't this resolved? What are the difficulties? When is a PR accepted? 

> does anyone feel like producing a small repo to reproduce this problem?:)

Does anyone has a small repo? Please post it instead of +1 ;)

**Comment by MrOrz** [comment]
For those who
- Uses webpack v3
- Hits this issue only in production builds (with extract-text-webpack-plugin)

Please upgrade your extract-text-webpack-plugin to 3.0.0, which is released just 6 hours ago. It solves my issue.

Related issue: https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/548

**Comment by vict-shevchenko** [comment]
I am not using a extract-text-webpack-plugin at all. I`ll be good to target the issue at its root, not by wrapping it with fixable plugin. I`ll try to provide a repo soon.

**Comment by CoolGoose** [comment]
Unfortunately it's still an issue for us with extract-text-webpack-plugin, any way of helping to debug this even further ?

**Comment by pim013** [comment]
+1 Please fix!

**Comment by julmot** [comment]
I have installed extract-text-wepback-plugin `^3.0.0` and still experiencing this issue.

**Comment by jpzwarte** [comment]
My problem was mixing ```require``` and ```import```: the former just for loading the styles, the latter for the ES6 modules. Changing the former to use the latter fixed my problem.

**Comment by GGAlanSmithee** [comment]
to add to @jpzwarte's comment, for those who didn't know, you can do `import "module-name"` which is equivalent to `require("module-name")` if you want anonymous imports.

**Comment by tremby** [comment]
I have this issue too.

- I'm not using extract text plugin.
- I'm hot loading.
- I'm using css-next.
- The problem scripts are normalize.css and my global.css. It doesn't matter what order I put them in, whether they're import or require or either combination. global.css always loads in the document before normalize.css.

I have found no solution other than to add specificity to the rules in global.css, which is obviously not good.

**Comment by jch254** [comment]
+1 Please fix!

**Comment by Peterabsolon** [comment]
I'm experiencing the same issue as others using the latest extract text plugin.

**Comment by 304NotModified** [comment]
timse commented on Mar 18
> does anyone feel like producing a small repo to reproduce this problem?:)

**Comment by julmot** [comment]
Especially @304NotModified 

https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/200#issuecomment-328675643

**Comment by josser** [comment]
@timse 
I've created a small repo to demo this issue
https://github.com/josser/css-order-example

However, I'm not sure this is correct example ))
@everyone guys, can you please check the code?

**Comment by Kovensky** [comment]
@josser what exactly is that supposed to demonstrate?

I haven't tried executing it, but from looking at the source, it looks like your `styles.scss` will be added to the `head` _before_ the material UI styles. This is because the material UI styles are added by `material-ui` at runtime on `componentDidMount` (styles done as css-in-js). There is nothing webpack can do to make material UI's styles be lower priority than your own styles.

**Comment by josser** [comment]
@Kovensky You answered your own question )
But I didn't know how css-in-js working. 
However if webpack can't resolve this, who can tho?

Maybe there is possibility to specify place for webpack css, so I can put them after material UI?

**Comment by GGAlanSmithee** [comment]
@josser in the case of [specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity), the order of inclusion does not matter. Make your own style have higher specificity somehow.

**Comment by Kovensky** [comment]
@GGAlanSmithee it does matter (if the calculated weight is the same, styles later in tree order, and if in the same tree order, later in text order have higher specificity). However, you can always boost your own specificity by adding a `:root` in front of your selector (`:root .button`); it'll still always match, but it'll have higher calculated weight.

**Comment by GGAlanSmithee** [comment]
> @GGAlanSmithee it does matter as it disambiguates styles that have the same weight

@Kovensky you are right. I was only referring to rules of higher specificity, not the same:

> in the case of __specificity__

Maybe I am mixing the terminology, but I guess I was talking about weight then, so let me rephrase:

If a rule with a higher weight precedes a conflicting rule with a lesser weight, the first rule will get applied.

Thanks for keeping things clear @Kovensky

**Comment by josser** [comment]
@GGAlanSmithee Can you please take a look at my repo, if I'm importing scss, I just have class name and nothing to control specificity

**Comment by delmicio** [comment]
> gunta opened this issue on 28 Mar 2014

really?

**Comment by speedDeveloper** [comment]
gatsbyjs/gatsby#1994

**Comment by Miller547719886** [comment]
so how do you handle the css order problem?

**Comment by AoDev** [comment]
@Miller547719886 Only one file required is what I've decided to do. All the imports are handled by my CSS pre-processor and I have 100% control about the order in which the styles are defined. I have an index.less (or scss, etc) which is the only one imported through Webpack.

**Comment by Miller547719886** [comment]
@AoDev what if I need to import the css in node_modules？ In my project，I need to import a UI Toolkit with its css file.In this case how should I handle the css file in production mode?I don't want to copy one and use it as my static assets.

**Comment by AoDev** [comment]
@Miller547719886 I import such files like this (with `~`)

Example:
```less
@import '~react-grid-layout/css/styles.css';
@import '~react-resizable/css/styles.css';
```

I try to avoid UI components / libraries that inject their style themselves.

**Comment by vict-shevchenko** [comment]
@Miller547719886 I personally have included extract-text plugin and configured it to produce 2 css files. `app.css`(general styles) and `components.css`(specific styles). The order they are included is defined in `index.html`.

Probably soon will switch to css-in-js solution.

**Comment by lancee** [comment]
+1 please fix, thanks.

**Comment by lancee** [comment]
Aha, I found the solution…

Change: `require("foo.css")` to `import "foo.css"` in JS, then I got the right order:

```
<style type="text/css">
/* global style */
</style>

<style type="text/css">
/* components style */
</style>
```

**Comment by AO17** [comment]
is it possible to extract Bootstrap and other libraries to it's own file, so it does not appear in inline-js in our code?

**Comment by luojianet** [comment]
same issue https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/200

**Comment by tedfitzpatrick** [comment]
"css-loader": "^1.0.0, "style-loader": "^0.21.0" -- When I used import() for a.css and b.css, b was injected into the html before a was injected. I changed import() to require(), and now the injection order matched the order of the requires

**Comment by TheDutchCoder** [comment]
This is still a problem, I'm using v3, but requiring Sass files does not maintain order in webpack.

It's very frustrating because order matters with CSS.

**Comment by tremby** [comment]
@tedfitzpatrick's suggestion of using `require` instead of `import` is not working for me. It's not practical for me to use a preprocessor for these stylesheets, so what options am I left with?

**Comment by CatBakun** [comment]
using `require` instead of `import` did not work for me.

**Comment by x-yuri** [comment]
I can't reproduce the issue. I tried with Webpack 3, 4, Bootstrap 3, 4, CSS, SASS, minification, source maps, production, development mode. If you still think it exists, make a PR to my [test repo][1], or at least tell me what I'm missing.

[1]: https://github.com/x-yuri/example-webpack-css-order

The only nuance here is that `imports` are hoisted. So if you mix `imports`s and `require`s, like,

```js
require('./2.css');
import './1.css';
```

Then `1.css` comes before `2.css`.

**Comment by vladimir-kondratenko** [comment]
Seriously, this issue continues for four years? This is very sad

**Comment by x-yuri** [comment]
@vladimir-kondratenko Try to read the thread more carefully next time.

**Comment by vladimir-kondratenko** [comment]
@x-yuri sry, what solution I've missed?

**Comment by x-yuri** [comment]
@vladimir-kondratenko The issue is that it's not clear how to reproduce the issue. Can you please [help us][1]? Do you not see a way to affect the order of CSS files? Or you have a better idea of how it's supposed to work?

[1]: https://github.com/webpack/webpack/issues/215#issuecomment-422092704

**Comment by egemon** [comment]
@x-yuri I also probably missing the same as @vladimir-kondratenko . I switched off ExtractTextWebpackPlugin and now order is correct, but with that plugin order is incorrect. I am using wepack v3. Could u tell me how to resolve issue with extract text plugin?

**Comment by x-yuri** [comment]
> Could u tell me how to resolve issue with extract text plugin?

@egemon Sure, if you give me repository or archive with files that reproduce the issue. The less code, the better.

**Comment by thomashibbard** [comment]
This is insanity this is still a thing. Also:

> And the module id depends on how often the id is used and alphabetical order. Or is random in dev mode.

While this seems to be true, it is very puzzling why it's the case. What good is a development environment if it doesn't even seek to emulate a production one? And worse, you have to find a four year old comment to understand this.

**Comment by x-yuri** [comment]
@thomashibbard I admit that webpack is no simple piece of software, but I'd say the order is more or less obvious or expected, granted you have an idea how webpack works. From the first page of [the documentation][1]:

> When webpack processes your application, it internally builds a dependency graph which maps every module your project needs and generates one or more bundles.

[1]: https://webpack.js.org/concepts/

You don't tell weback about the order in any way, so dependency graph basically determines the order.

> What good is a development environment if it doesn't even seek to emulate a production one?

The order must not change from one environment to another. Can you help to reproduce it?

Also, if you don't like its idea of order, what's yours? How would you like it to work?

**Comment by davisriska** [comment]
> > Could u tell me how to resolve issue with extract text plugin?
> 
> @egemon Sure, if you give me a way to reproduce it (repository or archive with the files needed to reproduce the issue).

Hello. I looked at your repository and it seems you are importing them inside a js file. Im not sure if that`s why its working for you, but for me its like this - 

I have app.scss file with -

`
@import '~bootstrap/scss/bootstrap';
@import "~bootstrap-social/bootstrap-social.css";
`

And bootstrap seems to be added after bootstrap-social. Maybe this helps you to figure this out? Im using laravel-mix. I understand that adds another element in the equation, but as I dont know much about the workings of webpack, this is what I use to make it easy.

**Comment by evilebottnawi** [comment]
The main problem with css order - css is not deterministic. Example:
`A.js`:
```js
@import 'foo.css`;
@import 'bar.css`;

`B.js`:
```js
@import 'bar.css`;
@import 'foo.css`;
```

`App.js`:
```js
@import 'a.js';
@import 'b.js';
```

When we extract css, who should be first?

**Comment by CoolGoose** [comment]
@evilebottnawi from my point of view it would be great to have a flag to make that error out and force the same order everywhere, since if there's any shared css in those files, A and B will have different outputs

**Comment by evilebottnawi** [comment]
@CoolGoose we already do warning in mini-css-extract-plugin, do you have other case where you have invalid order and no warnings?

**Comment by x-yuri** [comment]
@evilebottnawi I wouldn't call that nondeterministic. I suppose nondeterministic is when you don't change anything, but the resulting order changes. What we have in the example though is contradicting demands.

I take it the resulting order is determined by the first time a file was required. And a warning is given when css files are required in different orders. Which means... case closed? :)

I'd like to check the last suggested case, but it's unlikely we'll learn anything new.

**Comment by evilebottnawi** [comment]
@x-yuri we can have more difficult cases, so need good algorithm for this, maybe with options about preferences and controlling (this should be implemented in `mini-css-wepback-plugin`) and in next iteration union with `style-loader` and `css-loader` and when moving as css support out of box in webpack

**Comment by rxall** [comment]
**TLDR;** 
I had two imports `@import 'b'` and `@import 'a.css'`, removing the .css extension from the import statement allowed me to import in the desired order.

**Detail**
`b` would always appear before `a` in the css output regardless of the order I imported them in.

> The syntax is the same as now - @import "your/path/to/the/file", without an extension after the file name. This will import your file directly. If you append *.css at the end, it will translate into the css rule @import url(...).

[https://stackoverflow.com/a/30279590](https://stackoverflow.com/a/30279590)

Removing the .css extension from the `a` import statement solved my issue. I haven't dug into the source to see if `@import url` is treated differently when ordering over plain `@import` but it worked in this case.

**Comment by gabsprates** [comment]
I don't know if it could be an important point but: is webpack's dependency graph created synchronously or asynchronously?

**Comment by Abhi0725** [comment]
2020 and we don't have a fix for this yet?

**Comment by x-yuri** [comment]
It's 2020 and people (like @Abhi0725) are still asking those kind of questions? Instead of putting effort into it. E.g. one can report his/her specific case and tell what he/she expects `webpack` to do (exactly). Or read the issue carefully to see that (most likely) `webpack` does as expected.

@evilebottnawi So you believe there are still cases where `webpack`'s behavior is suboptimal, and you keep this issue open for people to report those cases?

**Comment by alexanderpharwood** [comment]
@x-yuri I shall attempt to clarify what I believe other people are experiencing and what I myself am, in relation to how I should expect webpack to behave in this scenario.

Given the following code in the root App.js in my react application:
```
import 'semantic-ui-css/semantic.min.css';
import './assets/scss/layout.scss';
```
I should expect the rules from layout.scss to be included in my bundle _after_ the rules from semantic. This is not the case. However, when I change the code to this:

```
import 'semantic-ui-css/semantic.min.css';
import './assets/scss/z_layout.scss';
```
(note the 'z' before layout) the rules in my layout file _are_ rendered after the rules from semantic.

This is not how I personally should have expected webpack to behave. I should expect the bundle to reflect the order in which the files were requires / imported, rather than including them alphabetically. Hopefully that helps clear things up.

**Comment by x-yuri** [comment]
> @x-yuri I shall attempt to clarify what I believe other people are experiencing and what I myself am, in relation to how I should expect webpack to behave in this scenario.
> 
> Given the following code in the root App.js in my react application:
> 
> ```
> import 'semantic-ui-css/semantic.min.css';
> import './assets/scss/layout.scss';
> ```
>
> I should expect the rules from layout.scss to be included in my bundle after the rules from semantic. This is not the case. However, when I change the code to this:
> 
> ```
> import 'semantic-ui-css/semantic.min.css';
> import './assets/scss/z_layout.scss';
> ```
>
> (note the 'z' before layout) the rules in my layout file are rendered after the rules from semantic.
> 
> This is not how I personally should have expected webpack to behave. I should expect the bundle to reflect the order in which the files were requires / imported, rather than including them alphabetically. Hopefully that helps clear things up.

@alexanderharwood (hopefully I guessed the username right) The behavior you're describing is not correct. Can you please create a test repository that I can `git clone && npm i && webpack` to reproduce the issue? Then I can investigate the case.

**Comment by kidzen** [comment]
+1

**Comment by maurisrx** [comment]
This might help 

> I can't reproduce the issue. I tried with Webpack 3, 4, Bootstrap 3, 4, CSS, SASS, minification, source maps, production, development mode. If you still think it exists, make a PR to my [test repo](https://github.com/x-yuri/issues/tree/webpack-css-order), or at least tell me what I'm missing.
> 
> CSS modules are concatenated in the order of execution (requiring/importing). The only nuance here is that `imports` are [hoisted](https://stackoverflow.com/questions/29329662/are-es6-module-imports-hoisted) to the beginning of a module. That's what `babel` does as well. So if you mix `imports`s and `require`s in the same file, like,
> 
> ```js
> require('./2.css');
> import './1.css';
> ```
> 
> Then `1.css` comes before `2.css`.

In my case this was my issue. I mixed `require` and `import` but `import` files were always loaded first. The solution is to use the same `require` or `import` for all CSS files import.

**Comment by gabsprates** [comment]
> This might help
> 
> > I can't reproduce the issue. I tried with Webpack 3, 4, Bootstrap 3, 4, CSS, SASS, minification, source maps, production, development mode. If you still think it exists, make a PR to my [test repo](https://github.com/x-yuri/issues/tree/webpack-css-order), or at least tell me what I'm missing.
> > CSS modules are concatenated in the order of execution (requiring/importing). The only nuance here is that `imports` are [hoisted](https://stackoverflow.com/questions/29329662/are-es6-module-imports-hoisted) to the beginning of a module. That's what `babel` does as well. So if you mix `imports`s and `require`s in the same file, like,
> > ```js
> > require('./2.css');
> > import './1.css';
> > ```
> > 
> > 
> > Then `1.css` comes before `2.css`.
> 
> In my case this was my issue. I mixed `require` and `import` but `import` files were always loaded first. The solution is to use the same `require` or `import` for all CSS files import.

Yeah, but don't do that manually. Instead, use some tool like this: https://github.com/ggascoigne/prettier-plugin-import-sort

**Comment by x-yuri** [comment]
Does the tool allow exceptions? Like, sort alphabetically but `z.css` should go before `d.css`? If not, that's not relevant here. Most people come here with order dependencies. Although, the tool itself sounds interesting.

**Comment by SUXUMI** [comment]
7 years has been passed since this ticket was submitted and the issue still exists.

The solution what I found is to use third party plugins .scss files OR .css files only, do not use mix of them! In this case importing order is OK. But if I mix third party plugins' .css and .scss files, then .css files always goes first.

**Comment by alexander-akait** [comment]
We cannot guarantee CSS ordering between modules, because in some cases it is not possible, can you provide example the problem, I will say how to fix it/why you need to use other orders and say why it happens

**Comment by SUXUMI** [comment]
> We cannot guarantee CSS ordering between modules, because in some cases it is not possible, can you provide example the problem, I will say how to fix it/why you need to use other orders and say why it happens

Dear Alexander,
Thank you for your reply

I've a very simple app.scss code:

```
@import '~bootstrap/scss/bootstrap';
@import '~third-party/sample.min.css';
```

in output, via mix, third-party styles goes first after compilation.

If sample is *.scss , then it's ok but sometimes such files are not provided.

What would be the way out in this case?

**Comment by alexander-akait** [comment]
It is not webpack problem, try to compile scss without webpack, `@import` with `CSS` extensions keeping as is after SCSS compiler, then css-loader resolve them and import, https://sass-lang.com/documentation/at-rules/import#plain-css-imports, try to remove `css` extension

**Comment by SUXUMI** [comment]
Dear @alexander-akait thanks
but here https://github.com/JeffreyWay/laravel-mix/issues/429#issuecomment-281103135 was said the same.

anyway, thanks for your time. try to play around.

**Comment by alexander-akait** [comment]
No webpack issue here, you have:
```css
@import '~bootstrap/scss/bootstrap';
@import '~third-party/sample.min.css';
```

after sass (`node-sass`/`sass` (dart)) you will have

```css
@import '~third-party/sample.min.css';

// Just example of code
.bootstrap-classes { color: red; }
```

After `css-loader` you will have:
```css
// Just example of code
.bootstrap-classes { color: blue; }

// Just example of code
.bootstrap-classes { color: red; }
```

This is misleading, but in fact this is how sass works.

**Comment by johnnykrk** [comment]
In my case the styles in bootstrap.css was always at top of my custom styles scss.
To fix that I changed the loaders order in webpack.dev.config.js to have fisrt the test: /\.scss$/:
            {
                test: /\.scss$/,
                use: [
                    'style-loader', 'css-loader', 'sass-loader'
                ]
            },
            {
                test: /\.css$/,
                use: [
                    'style-loader', 'css-loader'
                ]
            },

**Comment by alexander-akait** [comment]
I think we can close it, if somebody have problems with order, please open an issue in https://github.com/webpack-contrib/css-loader with example of the problem, currently most of the problems are solved

**Comment by thoughtspile** [comment]
If you're seeing out-of-oreder CSS inside node_modules, head over here: https://github.com/webpack/webpack/issues/7094

I brought a repro!
